[stable/influxdb] Adds support for InfluxDB Enterprise

wgj opened the pull request:
Follows the newly introduced pattern of StatefulSets for pods, extending InfluxDB's
sts and allowing parameter for it's ReplicaSet, and adds a new sts for InfluxDB
"meta nodes" which hold distribution state for InfluxDB's high availability and
horizontal scaling.

Documentation for these new features has been added, but I can not stress enough
how important it is to follow, as Enterprise not only uses different container
images for meta containers, but also expects new container images for InfluxDB
(appropriately labeled `data` in Docker Hub) which in my experience was easy to
miss if a user is following one set of [instructions](https://docs.influxdata.com/enterprise_influxdb/v1.7/install-and-deploy/production_installation/data_node_installation/) over [another](https://hub.docker.com/_/influxdb).

A caveat of trade off of deploying Enterprise is an expectation of human
interaction when creating or scaling the cluster (`influxd-ctl` usage). This
makes Enterprise decidedly not "cloud native", but is less obnoxious when using
a StatefulSet with PVCs, and no pod auto-scaling, of which I suppose is unlikely
given InfluxData's approach to licensing.

Still forcing k8s operator involvement feels "bad", and I left in provisions to
improve this expectation in a later version, namely a headless service and using
`publishNotReadyAddresses`, such that a pod can use the Service to find
available `influxdb-meta` and `influxdb` pods to create/join an InfluxDB
cluster.

Signed-off-by: Weston Johnson <wgj@automox.com>

<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->
#### Is this a new chart
No.

#### What this PR does / why we need it:
Adds support for InfluxDB Enterprise, specifically HA clustering.

#### Which issue this PR fixes
  - fixes #18685

#### Special notes for your reviewer:

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

Commit:
Adds missing newline at end of file

Signed-off-by: Weston Johnson <wgj@automox.com>

Commit:
Adds support for InfluxDB Enterprise

Follows the newly introduced pattern of StatefulSets for pods, extending InfluxDB's
sts and allowing parameter for it's ReplicaSet, and adds a new sts for InfluxDB
"meta nodes" which hold distribution state for InfluxDB's high availability and
horizontal scaling.

Documentation for these new features has been added, but I can not stress enough
how important it is to follow, as Enterprise not only uses different container
images for meta containers, but also expects new container images for InfluxDB
(appropriately labeled `data` in Docker Hub) which in my experience was easy to
miss if a user is following one set of [instructions](https://docs.influxdata.com/enterprise_influxdb/v1.7/install-and-deploy/production_installation/data_node_installation/) over [another](https://hub.docker.com/_/influxdb).

A caveat of trade off of deploying Enterprise is an expectation of human
interaction when creating or scaling the cluster (`influxd-ctl` usage). This
makes Enterprise decidedly not "cloud native", but is less obnoxious when using
a StatefulSet with PVCs, and no pod auto-scaling, of which I suppose is unlikely
given InfluxData's approach to licensing.

Still forcing k8s operator involvement feels "bad", and I left in provisions to
improve this expectation in a later version, namely a headless service and using
`publishNotReadyAddresses`, such that a pod can use the Service to find
available `influxdb-meta` and `influxdb` pods to create/join an InfluxDB
cluster.

Signed-off-by: Weston Johnson <wgj@automox.com>

Commit:
Merge branch 'master' into influxdb/enterprise

naseemkullah commented on stable/influxdb/templates/statefulset.yaml:
```diff
@@ -58,9 +62,19 @@ spec:
         env:
 {{ toYaml .Values.env | indent 10 }}
+        {{- if .Values.enterprise.enabled }}
+          - name: INFLUXDB_HOSTNAME # Values.env's HOSTNAME isn't fundamentally different from $HOSTNAME, but this way weg get a distinguished name at runtime.
```
Not familiar with enterprise so not sure how/what this is used for, but would `"{{ template "influxdb.fullname" . }}-<stateful set replica number>` (Which is value for $HOSTNAME) not be sufficient?

wgj replied:
Sorry, "distinguished" wasn't a great fit here. I meant like a FQDN. InfluxDB Enterprise needs names to be resolvable.

naseemkullah commented on stable/influxdb/templates/meta-statefulset.yaml:
```diff
@@ -0,0 +1,132 @@
+{{ toYaml .Values.env | indent 10 }}
+          # Values.env's HOSTNAME isn't fundamentally different from $HOSTNAME, but this way we get a distinguished name for InfluxDB at runtime.
+          - name: INFLUXDB_HOSTNAME
+            value: "$(_HOSTNAME).{{ template "influxdb.fullname" . }}-meta"
```
Not familiar with enterprise so not sure how/what this is used for, but would `"{{ template "influxdb.fullname" . }}-meta-<stateful set replica number>` (Which is value for $HOSTNAME) not be sufficiently distinguished?

wgj replied:
Sorry, "distinguished" wasn't a great fit here. I meant like a FQDN. InfluxDB Enterprise needs names to be resolvable.

naseemkullah commented on stable/influxdb/values.yaml:
```diff
@@ -56,6 +56,29 @@ persistence:
+##
+enterprise:
+  enabled: false
+  licensekey: {}
```
Would be nice to add an option to provide the license key via a pre existing k8s secret

wgj replied:
That's a good idea. I'd like to leave as an improvement for a later PR, if that's okay with you.

naseemkullah commented on stable/influxdb/templates/meta-statefulset.yaml:
```diff
@@ -0,0 +1,132 @@
+      {{- end }}
+      {{- end }}
+  {{- end }}
+{{- end }}
```
nit: empty line at EOF

wgj replied:
Thanks!

naseemkullah commented on stable/influxdb/README.md:
```diff
@@ -65,6 +65,39 @@ $ helm install --name my-release -f values.yaml stable/influxdb
+[InfluxDB Enterprise](https://www.influxdata.com/products/influxdb-enterprise/) is a hardened version of the open source core InfluxDB that includes additional closed source features designed for production workloads, featuring high availability and horizontal scaling. InfluxDB Enterprise features require a InfluxDB Enterprise license.
+
+#### Configuration
+To enable InfluxDB Enterprise, set the following keys and values in a values file provided to Helm.
```
Feel free to provide a sample `enterprise-values.yaml` file as well!

wgj replied:
Is that preferred? I tend to lean away from multiple values files, as the ordering is non-intuitive and easy to break (and I'm unaware if merging `.env` will do what I want). 

Everything in `.enterprise` could be kept separate, but seeing as how that part is a neat and contained, I think it makes sense to keep colocated.

naseemkullah reviewed (commented):
Don't know much about enterprise, but I left a few comments if you wish to consider them.

Commit:
Merge remote-tracking branch 'origin/influxdb/enterprise' into influxdb/enterprise

# Conflicts:
#	stable/influxdb/templates/meta-config.yaml
#	stable/influxdb/templates/meta-service.yaml
#	stable/influxdb/templates/meta-statefulset.yaml

Aisuko commented on stable/influxdb/values.yaml:
```diff
@@ -56,6 +56,29 @@ persistence:
+  clusterSize: 4
+  meta:
+    image:
+      tag: meta
```
Is there could add the link about how to make choice for the Influxdb tag?

Aisuko commented on stable/influxdb/values.yaml:
```diff
@@ -56,6 +56,29 @@ persistence:
 
+## Deploy InfluxDB Enterprise - License required
+##
+enterprise:
```
I believe we could add the link about the information of Influxdb Enterprise.

Aisuko replied:
Thanks

Aisuko reviewed (commented):
Thanks for the new feature, there may some comments, please have a look.

Aisuko reviewed (changes requested):
I believe for the Statefulset we need to add the `headless` type of service, please see the link below https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#limitations

Commit:
Add reference links for InfluxDB Enterprise

Signed-off-by: Weston Johnson <wgj@automox.com>

Commit:
Bump version

Signed-off-by: Weston Johnson <wgj@automox.com>

Commit:
Remove bad whitespace that sneaked in

Signed-off-by: Weston Johnson <wgj@automox.com>

Commit:
Merge branch 'master' into influxdb/enterprise

Commit:
Hard code headless service for metadata

Metadata service is a StatefulSet, so it doesn't make sense to let users
to set the k8s service to anything but ClusterIP::None. This is true for
the StatefulSet for InfluxDB data pods, but that will be fixed in a
subsequent PR. When that happens, the Enterprise section of README.md
should be updated to remove references to ClusterIP::None.

Signed-off-by: Weston Johnson <wgj@automox.com>

Commit:
Merge remote-tracking branch 'origin/influxdb/enterprise' into influxdb/enterprise

Aisuko reviewed (changes requested):
Maybe there was something wrong. I mean the Statefulset need two services.  For example "meta-service.yaml" and "meta-service-headless.yaml" would be good. The first one above exposes the service for end-user, another one for Influxdb Statefulset.

Just like https://github.com/helm/charts/tree/master/stable/rabbitmq/templates/svc-headless.yaml and 
https://github.com/helm/charts/tree/master/stable/rabbitmq/templates/svc.yaml

The service for Influxdb Statefulset(not the enterprise version) will be named `service.yaml`, `service-headless.yaml` and this will be included in the next PR.

Thanks for your patient, sir.
